<a href="https://colab.research.google.com/github/seodalzzz/ESAA/blob/Python_Machine_Learning/Python_Machine_Learning_CHAP1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#__CHAP1 파이썬 기반의 머신러닝과 생태계 이해__

## __1. 머신러닝의 개념__

머신러닝(Machine Learning) : 애플리케이션을 수행하지 않고도 데이터를 기반으로 패턴을 학습하고 결과를 예측하는 알고리즘 기법

머신러닝은 데이터를 기반으로 숨겨진 패턴을 인지해 여러 문제를 해결한다. 머신러닝 알고리즘은 데이터를 기반으로 통계적인 신뢰도를 강화하고 예측 오류를 최소화하기 위한 다양한 수학적 기법을 적용해 데이터 내의 패턴을 스스로 인지뢰고 신뢰도 있는 예측 결과를 도출해 낸다.


###1) 머신러닝의 분류

지도학습/비지도학습/강화학습

* 지도학습 : 분류, 회귀, 추천시스템, 시각/음성감지/인지, 텍스트분석, NLP
* 비지도학습 : 클러스터링, 차원 축소, 강화학습

###2) 데이터 전쟁
머신러닝은 데이터에 매우 의존적임. 따라서 좋은 품직의 데이터를 갖추지 못한다면 머신런이의 수행 결과도 좋을 수 없음.

데이터를 이해하고 효율적으로 가공, 처리, 추출해 최적의 데이터를 기반으로 알고리즘을 구동할 수 있도록 준비하는 능력이 중요함.


###3) 파이썬과 R 기반의 머신러닝 비교

머신러닝 프로그램을 작성할 수 있는 대표적인 오픈 소스 프로그램 언어

  * R
   * 통계 전용 언어
   * 오랜 기간동안 많은 R 사용자들이 생성하고 검증해온 다양하고 많은 통계 패키지 보유

  * Python
    * 쉽고 뛰어난 개발 생산성으로 전 세계 개발자들이 파이썬을 선호함
    * 오픈 소스 계열의 전폭적인 지지를 받고 있으며 많은 라이브러리로 인해 개발 시 높은 생산성을 보장해 줌
    * 인터프리터 언어(Interpreter Language)의 특성 상 속도는 느리지만, 뛰어난 확장성, 유연성, 호환성으로 인해 서버, 네트워크, 시스템, IOT, 데스크톱까지 다양한 영역에서 사용되고 있음
    * 머신러닝 애플리케이션과 결합한 다양한 애플리케이션 개발이 가능함
    * 엔터프라이즈 아키미처로의 확장 및 마이크로서비스 기반의 실시간 연계 등 다양한 기업 환경으로의 확산이 가능함
    * __딥러닝 프레임워크인 텐서플로, 케라스, 파이토치 등에서 파이썬 우선 전책으로 파이썬을 지원하고 있음__





##__2. 파이썬 머신러닝 생태계를 구성하는 주요 패키지__
  * 머신러닝 패키지 : __사이킷런(Scikit-Learn)__, 텐서플로, 케라스
  * 행렬/선형대수/통계 패키지 : __넘파이(Numpy)__
    + 사이킷런을 비롯한 많은 머신러닝 패키지가 넘파이 기반으로되 되어 있음
  * 데이터 핸들링 : __판다스__
    + 대표적인 데이터 처리 패키지
    + 2차원 데이터 처리에 특화되어 있음
  * 시각화 : __맷플롯립__, 시본
    + 시본을 사용하기 위해서는 맷플롯립을 어느정도 알고 있어야 함
    
  * 주피터 노트북 : 대표적인 아이파이썬(대화형 파이썬 툴) 지원 툴
    + 중요 코드 단위로 설명ㅇ르 적고 코드를 수행해 그 결과를 볼수 있게 만들어 직관적으로 어떤 코드가 어떤 역할을 하는지 매우 쉽게 이해할 수 있도록 지원함
  




##__3. 넘파이__
 머신러닝의 주요 알고리즘은 선형대수와 통계 등에 기반함.

 Numpy(Numerical Python)는 파이썬에서 선형대수 기반의 프로그램을 쉽게 만들 수 있도록 지원하는 대표적인 패키지

 루프를 사용하징 않고 대량 데이터의 배열 연산을 가능하게 함. >> 빠른 배열 연산 속도

 C/C++과 같은 저수준 언어 기반의 호환 API를 제공함으로 수행 성능이 중요한 부분은 C/C++ 기반의 코드로 작성하고, 이를 넘파이에서 호출하는 방식으로 쉽게 통합 가능

 다양한 데이터 핸들링 기능을 제공하지만 2차원 형태의 행렬 데이터에 대한 다양한 가공과 변환, 여러 통계용 함수의 적용 등에서 판다스의 편리성에는 미치지 못함

__많은 머신러닝 알고리즘이 넘파이 기반으로 작성돼 있고, 이들 알고리즘의 입력 데이터와 출력 데이터를 넘파이 배열 타입으로 사용하기때문에 넘파이를 이해하는 것은 파이썬 기반의 머신러닝에서 매우 중요함.__





###1) 넘파이 ndarry 개요


In [ ]:
# 넘파이 모듈 임포트

import numpy as np

> import numpy만 해도 충분하지만, as np를 추가하여 약어로 모듈을 표현해주는 것이 관례임




* ndarry
  + 넘파이의 기반 데이터 타입
  + ndarry를 이용해 넘파이에서 다차원 배열을 쉽게 생성하고 다양한 연산을 수행할 수 있음




+ array() : 파이썬의 리스트와 같은 다양한 인자를 입력받아서 ndarray로 변환
  + 생성된 ndarray 배열의 shape 변수는 ndarray의 크기, 즉 행과 열의 수를 튜플 평태로 가지고 있으며 이를 통해 ndarray 배열의 차원까지 알 수 있음
  + array() 함수의 인자로는 파이썬의 리스트 객체가 주로 사용됨. 리스는 []는 1차원이고, [[]]는 2차원과 같은 형태로 배열의 차원과 크기를 쉽게 표현할 수 있기 때문








In [ ]:
# aray() 함수

array1=np.array([1,2,3]) #1차원 array로 3개의 데이터를 가지고 있음
print('array1 type : ',type(array1))
print('array1 array 형태 :',array1.shape)

array2=np.array([[1,2,3],[2,3,4]])
print('array2 type : ', type(array2))
print('arry2 형태 : ',array2.shape)

array3=np.array([[1,2,3]]) #1개의 row와 3개의 col으로 구성된 2차원 데이터
print('array3 type : ', type(array3))
print('array3 형태 : ', array3.shape)


array1 type :  <class 'numpy.ndarray'>
array1 array 형태 : (3,)
array2 type :  <class 'numpy.ndarray'>
arry2 형태 :  (2, 3)
array3 type :  <class 'numpy.ndarray'>
array3 형태 :  (1, 3)





> array1.shape // (3,) : 1차원 array로 3개의 데이터를 가지고 있음

> array3.shape // (1,3) : 1개의 row, 3개의 col을 가진 2차원 데이터






In [ ]:
# 각 array의 차원을 ndarray.ndim을 이용해 확인

print('array1 : {:0}차원, array2 : {:1}차원, array3 : {:2}차원'.format(array1.ndim,array2.ndim,array3.ndim))

array1 : 1차원, array2 : 2차원, array3 :  2차원




> array1은 1차원, array3은 2차원




###2) ndarray의 데이터 타입

  + ndarray내의 데이터 값은 숫자 값, 문자열 값, bool 값 등이 모두 가능함
  + 숫자형의 경우 int 형, unsigned int 형, float 형, complex 타입 제공
  + ndarray내의 데이터 타입은 같은 데이터 타입만 가능함
    + 만약 다른 데이터 유형이 섞여 있는 리스트를 ndarray로 변경하면 데이터 크기가 더 큰 데이터 타입으로 형 변환을 일괄 적용함



In [ ]:
# dtype 속성으로 ndarray 내의 데이터 타입 확인하기
list1=[1,2,3]
print(type(list1))
array1=np.array(list1)
print(type(array1))
print(array1,array1.dtype)

<class 'list'>
<class 'numpy.ndarray'>
[1 2 3] int64




> list1을 array()함수를 이용하여 ndarray로 변환
> 변경된 ndarray내의 데이터 값은 모두 int32형




In [ ]:
#int형과 string형이 섞여 있는 리스트를 ndarray로 변경
list2=[1,2,'test']
array2=np.array(list2)
print(array2, array2.dtype)

#int형과 float형이 섞여 있는 리스트를 ndarray로 변경
list3=[1,2,3.0]
array3=np.array(list3)
print(array3,array3.dtype)


['1' '2' 'test'] <U21
[1. 2. 3.] float64




> int 형 값과 문자열이 섞여 있는 list2를 ndarray로 변환하면 숫자형 값 1,2,가 모두 문자열 '1','2'로 바뀜
> int 형과 float형이 섞인 list3의 경우도 int 1,2가 모두 1,2 인 float 형으로 바뀜




* astype() : ndarray 내 데이터 값의 타입 변경 
  + astype()에 인자로 원하는 타입을 문자열로 지정
  + 대용량 데이터의 ndarray를 만들 때 많은 메모리가 사용되는데, 메모리를 절약해야 할 때 사용 (ex. int 형으로 충분한 경우인데 float형이 사용되고 있는 경우)


In [ ]:
# int32형 데이터를 float64로 변환
array_int=np.array([1,2,3])
array_float=array_int.astype('float64')
print(array_float,array_float.dtype)

#float64형 ndarray를 다시 int32형으로 변경
array_int1=array_float.astype('int32')
print(array_int1,array_int1.dtype)

array_float1=np.array([1.1,2.1,3.1])
array_int2=array_float1.astype('int32')
print(array_int2,array_int2.dtype)

[1. 2. 3.] float64
[1 2 3] int32
[1 2 3] int32


> float형을 int형으로 변환 할 때 소수점 이하의 자리는 모두 없어짐


###3) ndarray를 편리하게 생성하기 - arange, zeros, ones
  * 특정 크기와 차원을 가진 ndarray를 연속값이나 0 또는 1로 초기화해 생성해야 할 경우 사용
  * 테스트용 데이터를 만들거나 대규모의 데이터를 일괄적으로 초기화해야 할 경우에 사용

  * arange() : 0부터 함수인자 값 -1까지의 값을 순차적으로 ndarray의 데이터값으로 변환
    + default 함수 인자 값은 stop 값이며 0부터 stop 값에 -1을 더한 값까지의 연속 숫자 값으로 구성된 1차원 array를 만듦
    + range와 유사하게 start값도 부여해 0이 아닌 다른 값부터 시작한 연속 값을 부여할수도 있음

  * zeros() : 함수 인자로 튜플 형태의 shape값을 입력하면 모든 값을 0으로 채운 해당 shape를 가진 ndarray를 반환
    + 함수인자로 dtype을 정해주지 않으면 default로 float64 형의 데이터로 ndarray를 채움

  * ones() : 함수인자로 튜플 형태의 shape 값을 입력하면 모든 값을 1로 채운 해당 shape를 가진 ndarray를 반환
    + 함수인자로 dtype을 정해주지 않으면 default로 float64 형의 데이터로 ndarray를 채움

    


In [ ]:
#arange()
sequence_array=np.arange(10)
print(sequence_array)
print(sequence_array.dtype, sequence_array.shape)

#zeros()
zero_array=np.zeros((3,2),dtype='int32')
print(zero_array)
print(zero_array.dtype, zero_array)
  
#ones()
one_array=np.ones((3,2))
print(one_array)
print(one_array.dtype,one_array.shape)

[0 1 2 3 4 5 6 7 8 9]
int64 (10,)
[[0 0]
 [0 0]
 [0 0]]
int32 [[0 0]
 [0 0]
 [0 0]]
[[1. 1.]
 [1. 1.]
 [1. 1.]]
float64 (3, 2)


###4) ndarray의 차원과 크기를 변경하는 reshape()

  * reshape() : ndarray를 특정 차원 및 크기로 변환
    * 변환을 원하는 크기를 함수 인자로 부여
    * 지정된 사이즈로 변경이 불가능하면 오류 발생
    * 인자로 -1을 사용할 경우 원래 ndarray와 호환되는 새로운 shape로 변환

In [ ]:
#reshape()
array1=np.arange(10)
print('array1:\n',array1)

array2=array1.reshape(2,5)
print('array2:\n',array2)

array3=array1.reshape(5,2)
print('array3:\n',array3)

array1:
 [0 1 2 3 4 5 6 7 8 9]
array2:
 [[0 1 2 3 4]
 [5 6 7 8 9]]
array3:
 [[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]




> arrya1을 (4,3) 형태로 reshape하려 하면 오류 발생



In [ ]:
#reshape 인자로 -1 사용
array1=np.arange(10)
print(array1)
array2=array1.reshape(-1,5)
print('array2 shape:',array2.shape)
array3=array1.reshape(5,-1)
print('array3.shape:',array3.shape)


[0 1 2 3 4 5 6 7 8 9]
array2 shape: (2, 5)
array3.shape: (5, 2)


In [ ]:
#호환되지 않는 형태로는 변환 불가
array1=np.arange(10)
array4=array1.reshaep(-1,4)

AttributeError: ignored

In [ ]:
#reshape(-1,1)
array1=np.arange(8)
array3d=array1.reshape((2,2,2))
print('array3d:\n',array3d.tolist())

#3차원 ndarray를 2차원 ndarray로 변환
array5=array3d.reshape(-1,1)
print('array5:\n', array5.tolist())
print('array5 shape:',array5.shape)

#1차원 ndarray를 2차원 ndarray로 변환
array6=array1.reshape(-1,1)
print('array6:\n',array6.tolist())
print('array6 shape:\n',array6.shape)

array3d:
 [[[0, 1], [2, 3]], [[4, 5], [6, 7]]]
array5:
 [[0], [1], [2], [3], [4], [5], [6], [7]]
array5 shape: (8, 1)
array6:
 [[0], [1], [2], [3], [4], [5], [6], [7]]
array6 shape:
 (8, 1)


> -1인자는 reshape(-1,1)과 같은 형태로 자주사용됨. reshape(-1,1)은 원본 ndarray가 어떤 형태라도 2차원이고, 여러개의 로우를 가지되 반드시 1개의 칼럼을 가진 ndarray로 변환됨을 보장

> 여러 개의 넘파이 adarray는 stack이나 concat으로 결합할 때 각각의 ndarray 형태를 통일해 유용하게 사용됨

###5) 넘파이의 ndarray의 데이터 세트 선택하기 - 인덱싱(Indexing)
  * 넘파이에서 ndarray 내의 일부 데이터 세트나 특정 데이터만을 선택
  * 특정한 데이터만 추출 : 원하는 위치의 인덱스 값을 지정하면 해당 위치의 데이터가 반환
  * 슬라이싱(Slicing) : 연속된 인덱스상의 ndarray를 추출
    * :기호 사이에 시작 인덱스와 종료 인덱스를 표시하면 시작 인덱스에서 종료 인덱스 -1 위치에 있는 데이터의 ndarray를 반환
  * 팬시 인덱싱(Fancy Indexing) : 일정한 인덱싱 집합을 리스트 또는 ndarray 형태로 지정해 해당 위치에 있는 데이터의 ndarray를 반환
  * 불린 인덱싱 (Boolean Indexing) : 특정 조건에 해당하는지 여부인 True/False 값 인덱싱 집합을 기반으로 True에 해당하는 인덱스 위치에 있는 데이터의 ndarray를 반환



####(1) 단일 값 추출
  * 1개의 데이터값을 선택하려면 ndarray 객체에 해당하는 위치의 인덱스 값을 []안에 입력하면 됨
  * 인덱스에 마이너스 기호를 이용하면 맨 뒤에서부터 데이터를 추출할 수 있음

In [ ]:
#단일 값 추출

#1부터 9까지의 1차원 ndarray 생성
array1=np.arange(1,10)
print('array1:',array1)

#index는 0부터 시작하므로 array1[2]는 세번째 index 위치의 데이터값을 의미
value=array1[2]
print('value:',value)
print(type(value))

array1: [1 2 3 4 5 6 7 8 9]
value: 3
<class 'numpy.int64'>


> array1[2] 의 값은 더이상 ndarray 타입이 아니고 ndarray내의 데이터 값을 의미함


In [ ]:
#인덱스에 마이너스 값 이용
print('맨뒤의 값 : ', array1[-1],'맨 뒤에서 두번째 값 : ',array1[-2])

맨뒤의 값 :  9 맨 뒤에서 두번째 값 :  8


In [ ]:
#인덱스를 이용해 ndarray내의 데이터 값 수정하기
array1[0]=9
array1[8]=0
print('array1:',array1)

array1: [9 2 3 4 5 6 7 8 0]


In [ ]:
#다차원 ndarray에서 단일 값 추출
array1d=np.arange(start=1,stop=10)
array2d=array1d.reshape(3,3)
print(array2d)

print('(row=0, col=0) index 가리키는 값 : ',array2d[0,0])
print('(row=0, col=1) index 가리키는 값 : ',array2d[0,1])
print('(row=1, col=0) index 가리키는 값 : ',array2d[1,0])
print('(row=2, col=2) index 가리키는 값 : ',array2d[2,2])

[[1 2 3]
 [4 5 6]
 [7 8 9]]
(row=0, col=0) index 가리키는 값 :  1
(row=0, col=1) index 가리키는 값 :  2
(row=1, col=0) index 가리키는 값 :  4
(row=2, col=2) index 가리키는 값 :  9


> array2d[0,1]은 첫번째 로우, 두번째 칼럼 위치의 데이터인 2


  * [row=0, col=1] 인덱싱은 [axis0=0, axis1=1]이 정확한 표현임
    * 일반적인 로우, 칼럼 형식은 넘파이에서는 사용되지 않음
    * 다차원 ndarray의 경우 축(axis)에 따른 연산을 지원함
    * 축 기반의 연산에서 axis가 생략되면 axis 0을 의미함


####(2) 슬라이싱
  * ':' 기호를 이용해 연속한 데이터를 슬라이싱해서 추출
  * ':' 기호 사이에 시작 인덱스와 종료 인덱스를 표시하면 시작 인텍스에서 종료 인덱스 -1의 위치에 있는 데이터의 ndarray 값을 반환
  * 단일 데이터값 추출을 제외하고 슬라이싱, 팬시 인덱싱, 불린 인덱싱으로 추출된 데이터 세트는 모두 ndarray 타입



In [ ]:
# 슬라이싱
array1=np.arange(start=1,stop=10)
array3=array1[0:3]
print(array3)
print(type(array3))


[1 2 3]
<class 'numpy.ndarray'>


  * 슬라이싱 기호인 ':'사이의 시작, 종료 인덱스는 생략 가능
    * 시작 인덱스 생략 > 맨 처음 인덱스인 0으로 간주
    * 종료 인덱스 생략 > 맨 마지막 인덱스로 간주
    * ':' 기호 앞 뒤에 시작/종료 인덱스 생략 > 맨 처음/맨 마지막 인덱스로 간주

In [ ]:
# ':' 기호 사용하여 시작 인덱스와 종료 인덱스 설정
array1=np.arange(start=1,stop=10)
array4=array1[:3]
print(array4)

array5=array1[3:]
print(array5)

array6=array1[:]
print(array6)


[1 2 3]
[4 5 6 7 8 9]
[1 2 3 4 5 6 7 8 9]


In [ ]:
# 2차원 ndarray에서의 슬라이싱
array1d=np.arange(start=1,stop=10)
array2d=array1d.reshape(3,3)
print('array2d:n',array2d)

print('array2d[0:2,0:2]\n',array2d[0:2,0:2])
print('array2d[1:3,0:3]\n',array2d[1:3,0:3])
print('array2d[1:3,:]\n',array2d[1:3,:])
print('array2d[:,:]\n',array2d[:,:])
print('array2d[:2,1:]\n',array2d[:2,1:])
print('array2d[:2,0]\n',array2d[:2,0])

array2d:n [[1 2 3]
 [4 5 6]
 [7 8 9]]
array2d[0:2,0:2]
 [[1 2]
 [4 5]]
array2d[1:3,0:3]
 [[4 5 6]
 [7 8 9]]
array2d[1:3,:]
 [[4 5 6]
 [7 8 9]]
array2d[:,:]
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
array2d[:2,1:]
 [[2 3]
 [5 6]]
array2d[:2,0]
 [1 4]


> array[:2,0]과 같이 로우나 칼럼 축 한쪽에만 슬라이싱을 적용하고, 다른 쪽 축에는 단일 값 인덱스를 적용해도 됨

  * 2차원 ndarray에서 뒤에 오는 인덱스를 없애면 1차원 ndarray를 반환 

    (ex.array2d[0]과 같이 2차원에서 뒤에 오는 인덱스를 없애면 로우 축(axis0)의 첫번째 로우 ndarray)를 반환)

In [ ]:
# n차원 ndarray에서 마지막 인덱스 지정 생략
print(array2d[0])
print(array2d[1])
print('array2d[0] shape : ', array2d[0].shape,'array2d[1] shape : ',array2d[1].shape)

[1 2 3]
[4 5 6]
array2d[0] shape :  (3,) array2d[1] shape :  (3,)


####(3) 팬시 인덱싱
  * 리스트나 ndarray로 인덱스 집합을 지정하면 해당 위치의 인덱스에 해당하는 ndarray를 반환하는 인덱싱 방식


In [ ]:
# 팬시 인덱싱
array1d=np.arange(start=1,stop=10)
array2d=array1d.reshape(3,3)

array3=array2d[[0,1],2]
print('array2d[[0,1],2] => ', array3.tolist())

array4=array2d[[0,1],0:2]
print('array2d[[0:1],0:2] => ',array4.tolist())

array5=array2d[[0,1]]
print('array2d[[0,1]] => ',array5.tolist())

array2d[[0,1],2] =>  [3, 6]
array2d[[0:1],0:2] =>  [[1, 2], [4, 5]]
array2d[[0,1]] =>  [[1, 2, 3], [4, 5, 6]]


> array2d[[0,1],2] 로우 축에 팬시 인덱싱인 [0,1]을, 칼럼 축에는 단일 값 인덱싱 2를 적용

> array2d[[0,1],0:2]는 ((0,0),(0,1)),((1,0),(1,1))이 적용

####(4) 불린 인덱싱
  * 조건 필터링과 검색 가능
  * for loop / if else 문을 간단하게 구현 가능
  *ndarray의 인덱스를 지정하는 []내에 조건문을 그대로 기재하기만 하면 됨

In [ ]:
# 불린 인덱싱
array1d=np.arange(start=1,stop=10)
# []안에 array1d>5 Boolean indexing을 적용
array3=array1d[array1d>5]
print('array1d > 5 불린 인덱싱 결과 값 : ', array3)

array1d > 5 불린 인덱싱 결과 값 :  [6 7 8 9]


In [ ]:
# 불린 인덱싱의 조건 필터링 과정
array1d>5

boolean_indexes=np.array([False,False,False,False,False,True,True,True,True])
array3=array1d[boolean_indexes]
print('불린 인덱스로 필터링 결과 : ',array3)

indexes=np.array([5,6,7,8])
array4=array1d[indexes]
print('일반 인덱스로 필터링 결과 : ',array4)

불린 인덱스로 필터링 결과 :  [6 7 8 9]
일반 인덱스로 필터링 결과 :  [6 7 8 9]


> 5보다 큰 데이터가 있는 위치는 True 값이, 그렇지 않은 경우는 False 값이 반환됨

> 조건으로 반환된 이 ndarray 객체를 인덱싱을 지정하는 []내에 입력하면 False값은 무시하고 True 값이 있는 위치 인덱스 값으로 자동 변환해 해당하는 인덱스 위치의 데이터만 반환

  * 불린 인덱싱이 동작하는 단계 순차적 도식화

     step1 : array1d>5 와 같이 ndarray의 필터링 조건을 [] 안에 기재

     step2 : False 값은 무시하고 True 값에 해당하는 인덱스값만 저장 (유의해야 할 사항은 True 값 자체인 1을 저장하는 것이 아니라 True 값을 가진 인덱스를 저장한다는 것)
     
     step3 : 저장된 인덱스 데이터 세트로 ndarray 조회 

### 6) 행렬의 정렬 - sort()와 argsort()

####(1) 행렬 정렬
  * np.sort() :  넘파이에서 sort()를 호출하는 방식
    * 원 행렬은 그대로 유지한 채 원 행렬의 정렬된 행렬을 반환
  
  * ndarray.sort() : 행렬 자체에서 sort()를 호출하는 방식
    * 원 행렬 자체를 정렬한 형태로 변환, 반환 값은 None

In [ ]:
org_array=np.array([3,1,9,5])
print('원본 행렬 : ',org_array)

#np.sort()로 정렬
sort_array1=np.sort(org_array)
print('np.sort() 호출 후 반환된 정렬 행렬 : ', sort_array1)
print('np.sort() 호출 후 원본 행렬 : ', org_array)

#ndarray.sort()로 정렬
sort_array2=org_array.sort()
print('org_array.sort() 호출 후 반환된 행렬 : ',sort_array2)
print('org_array.sort() 호출 후 원본 행렬 : ',org_array)


원본 행렬 :  [3 1 9 5]
np.sort() 호출 후 반환된 정렬 행렬 :  [1 3 5 9]
np.sort() 호출 후 원본 행렬 :  [3 1 9 5]
org_array.sort() 호출 후 반환된 행렬 :  None
org_array.sort() 호출 후 원본 행렬 :  [1 3 5 9]


> 원본 행렬에 대해서 np.sort()는 원본 행렬을 변경하지 않고 정렬된 형태로 반환했으며, ndarray.sort()는 원본 행렬 자체를 정렬한 값으로 변환함

  * np.sort()나 ndarray.sort() 모두 깁ㄴ적으로 오름차순으로 행렬 내 원소를 정렬함
  * 내림차순으로 정렬하기 위해서는 [::-1] 을 적용
  

In [ ]:
# 내림차순 정렬
sort_array1_desc=np.sort(org_array[::-1])
print('내림차순으로 정렬 : ',sort_array1_desc)

내림차순으로 정렬 :  [1 3 5 9]


  * 행렬이 2차원 이상일 경우에 axis 축 값 설정을 통해 로우방향, 또는 칼럼 방향으로 정렬을 수행할 수 있음

In [ ]:
array2d=np.array([[8,12],
                  [7,1]])

sort_array2d_axis0=np.sort(array2d,axis=0)
print('로우방향으로 정렬 : \n' , sort_array2d_axis0) 
#세로 방향으로 정렬, 같은 열에 있는 데이터들끼리 정렬

sort_array2d_axis1=np.sort(array2d,axis=1)
print('칼럼 방향으로 정렬 : \n', sort_array2d_axis1)
#가로 방향으로 정렬, 같은 행에 있는 데이터들끼리 정렬

로우방향으로 정렬 : 
 [[ 7  1]
 [ 8 12]]
칼럼 방향으로 정렬 : 
 [[ 8 12]
 [ 1  7]]


####(2) 정렬된 행렬의 인덱스를 반환하기

  * np.argsort() : 정렬된 행렬의 원본 행렬 인덱스를 ndarray 형으로 반환

    * 원본 행렬이 정렬되었을 때 기존 원본 행렬의 원소에 대한 인덱스를 피룡로 할 때 사용


In [ ]:
org_array=np.array([3,1,9,5])
sort_indices=np.argsort(org_array)
print(type(sort_indices))
print('행렬 정렬 시 원본 행렬의 인덱스 : ', sort_indices)

<class 'numpy.ndarray'>
행렬 정렬 시 원본 행렬의 인덱스 :  [1 0 3 2]


  * 내림차순으로 정렬 시에 원본 행렬의 인덱스를 구하는 것도 np.argsort()[::-1]과 같이 적용하면 됨
  

In [ ]:
# np.argsort()[::-1]
org_array=np.array([3,1,9,5])
sort_indices_desc=np.argsort(org_array)[::-1]
print('행렬 내림차순 정렬 시 원본 행렬의 인덱스 : ', sort_indices_desc)


행렬 내림차순 정렬 시 원본 행렬의 인덱스 :  [2 3 0 1]



  * 넘파이의 ndarray는  RDBMS 의 TABLE 칼럼이나 뒤에 소개할 판다스 DataFrame 칼럼과 같은 메타 데이터를 가질 수 없음. 따라서 실제 값과 그 값이 뜻하는 메타 데이터를 별도의 ndarray로 각각 가져야만 함
    * ex. 학생의 이름과 시험 성적을 ndarray 형태의 데이터로 가지려면 학생 ndarray와 성적 ndarray를 각각 가져야 함
    * 이 때 시험 성적순으로 학생 이름을 출력하고자 한다면 np.argsort(성적)을 이용해 반환된 인덱스를 학생 ndarray에 팬시 인덱스로 적용해 추출할 수 있음.

In [ ]:
import numpy as np

name_array=np.array(['John','Mike','Sarah','Kate','Samuel'])
score_array=np.array([78,95,84,98,88])

sort_indices_asc=np.argsort(score_array)
print('성적 오름차순 정렬 시 score_array의 인덱스 : ',sort_indices_asc)
print('성적 오름차순으로 name_array의 이름 출력 : ', name_array[sort_indices_asc])

성적 오름차순 정렬 시 score_array의 인덱스 :  [0 2 4 1 3]
성적 오름차순으로 name_array의 이름 출력 :  ['John' 'Sarah' 'Samuel' 'Mike' 'Kate']


###7) 선형대수 연산 - 행렬 내적과 전치 행렬 구하기

####(1) 행렬 내적(행렬 곱)
  * 행렬 내적 = 행렬 곱
  * np.dot()
  * 행렬의 내적은 왼쪽 행렬의 행과 오른쪽 행렬의 열의 원소들을 순차적으로 곱한 뒤 그 결과를 모두 더한 값
  * 왼쪽 행렬의 열 개수와 오른쪽 행렬의 행 개수가 동일해야 내적 연산이 가능함
  

In [ ]:
# 행렬 내적
A=np.array([[1,2,3],
            [4,5,6]])
B=np.array([[7,8],
            [9,10],
            [11,12]])

dot_product=np.dot(A,B)
print('행렬 내적 결과 : \n', dot_product)


행렬 내적 결과 : 
 [[ 58  64]
 [139 154]]


####(2) 전치 행렬
  * 원 행렬에서 행과 열 위치를 교환한 원소로 구성한 행렬
  * transpose()
  

In [ ]:
# 전치 행렬
A=np.array([[1,2],
            [3,4]])
transpose_mat=np.transpose(A)
print('A의 전치행렬 : \n',transpose_mat)

A의 전치행렬 : 
 [[1 3]
 [2 4]]


##**4. 데이터 핸들링 - 판다스**

행(row)과 열(col)로 이루어진 2차원 데이터를 효율적으로 가공/처리할 수 있는 다양한 기능을 제공함.

파이썬의 리스트, 컬렉션, 넘파이 등의 내부 데이터뿐만 아니라 CSV 등의 파일을 쉽게 DataFrame으로 변경해 데이터의 가공/분석을 편리하게 수행할 수 있게 만들어 줌.

DataFrame : 여러 개의 행과 열로 이뤄진 2차원 데이터를 담는 데이터 구조체

Series : 칼럼이 한인 데이터 구조체
  * DataFrame 은 여러 개의 Series로 이루어져 있음

Index : 개별 데이터를 고유하게 식별하는 Key 값
  * Series와 DataFrame 모두 Index를 Key 값으로 가짐




### 1) 판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API

In [8]:
import pandas as pd

> pandas를 pd로 에일리어스(alias)해 임포트하는 것이 관례임


* 캐글의 타이타닉 탑승자 데이터
  * https://www.kaggle.com/c/titanic/data
  * 맨 윗줄 칼럼명, 각 필드는 콤마로 분리되어 있음

* read_csv() : CSV 파일 포맷 변환을 위한 API
  * 호출 시 파일명 인자로 들어온 파일을 로딩해 DataFrame 객체로 반환
  * 별다른 파라미터 지정이 없으면 파일의 맨 처음 로우를 칼럼명으로 인지하고 칼럼으로 변환
  * 모든 DataFrame 내의 데이터는 생성되는 순간 고유의 index 값을 가지게 됨
  * sep : 구분 문자를 입력하여 해당 필드 구분 문자 기반의 파일 포맷도 DataFrame으로 변환 가능
    * ex. read_csv('파일명', sep='\t')
    * sep 인자 default는 ','
  * filepath : 로드하려는 데이터 파일의 경로를 포함한 파일명 입력
    * 파일명만 입력 시 파이썬 실행 파일이 있는 디렉터리와 동일한 디렉터리에 있는 파일명을 로딩
* read_table() : 필드 구분자가 탭('\t')인 파일 포맷 변환을 위한 API
* read_fwf() : Fixed Width, 즉 고정 길이 기반의 칼럼 포맷을 DataFrame으로 로딩하기 위한 API
* DataFrame.head()
  * DataFrame의 맨 앞에 있는 N(default 5)개의 로우를 반환함

In [10]:
titanic_df=pd.read_csv('titanic_train.csv')
titanic_df.head(3)


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [ ]:
print('titanic 변수 type : ',type(titanic_df))

titanic 변수 type :  <class 'pandas.core.frame.DataFrame'>


* dataFrame.shape : DataFrame의 행과 열을 튜플 형태로 반환

In [ ]:
print('DataFrame 크기 : ', titanic_df.shape)

DataFrame 크기 :  (891, 12)


> 생성된 DataFrame 객체인 titanic_df는 891개의 로우와 12개의 칼럼으로 이루어짐

* DataFrame.info() : 총 데이터 건수와 데이터 타입, Null 건수 확인 가능



In [ ]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


* DataFrame.describe() : 칼럼별 숫자형 데이터값의 n-percentile 분포도, 평균값, 최댓값, 최솟값을 나타냄
  * 숫자형(int,float 등) 칼럼의 분포도만 조사하며 자동으로 object 타입의 칼럼은 출력에서 제외시킴
  * count : Null이 아닌 데이터 건수
  * mean : 전체 데이터의 평균값
  * std : 표준편차
  * min : 최솟값
  * max : 최댓값
  * 25% : 25 percentile / 50% : 50 percentile / 75% : 75 percentile
  * 숫자 칼럼이 숫자형 카테고리 칼럼인지 판단하는 것을 도와 줌
    * 카테고리 칼럼 : 특정 범주에 속하는 값을 코드화한 칼럼
      * ex. 남/여 => 1,2

In [ ]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

> PassengerID : 승객 ID 식별된 칼럼으로 1~891까지 숫자가 할당되어 분석을 위한 의미있는 속성이 아님

> Survived : min 0, 25%~75% 0, max 1 => 0과 1로 이루어진 숫자형 카테고리 칼럼일 것


* DataFrame['Column'].value_counts()
  * Series 형태로 특정 칼럼 데이터 세트가 반환됨
    * Series : 하나의 칼럼으로 구성된 데이터 세트
  * value_counts() : 지정된 칼럼의 데이터값 건수를 반환
    * Series 객체에서만 정의되어 있음
    * 데이터의 분포도를 확인하는데 유용함
    * 건수가 많은 순서대로 정렬되어 Series 객체로 반환함


In [ ]:
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)


3    491
1    216
2    184
Name: Pclass, dtype: int64


> Pclass 값 3이 491개, 1이 216개, 2가 184개

In [ ]:
titanic_pclass=titanic_df['Pclass']
print(type(titanic_pclass))
titanic_pclass.head()

<class 'pandas.core.series.Series'>


0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

> 왼쪽의 0부터 시작하는 값 : 데이터프레임의 인덱스와 Index

> 오른쪽은 Series의 해당 칼럼의 데이터 값

In [ ]:
value_coutns=titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_coutns)

<class 'pandas.core.series.Series'>
3    491
1    216
2    184
Name: Pclass, dtype: int64


> 왼쪽 : Series의 인덱스 값 / 오른쪽 : Series의 데이터 값

> value_counts()는 칼럼 값별 데이터 건수를 반환하므로 고유 칼럼 값을 식별자로 사용할 수 있음

* Index
  * DataFrame, Series가 만들어진 후에도 변경 가능
  * 숫자형뿐만 아니라 문자열도 가능
  * 모든 인덱스는 고유성이 보장되어야 함

###2) DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환

기본적으로 DataFrame은 파이썬의 리스트, 딕셔너리, 넘파이 ndarray 등 다양한 데이터로부터 생성될 수 있으며 반대로 DataFrame 역시 파이썬의 리스트, 딕셔너리, 넘파이 ndarray 등으로 변환될 수 있음

####(1) 넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기
  * DataFrame은 리스트와 넘파이 ndarray와 다르게 칼럼명을 가지고 있어 핸들링이 편함
  * 일반적으로 DataFrame으로 변환 시 칼럼명을 지정 (지정하지 않으면 자동으로 칼럼명 할당)
  * DataFrame은 기본적으로 행과 열을 가지는 2차원 데이터이므로 2차원 이하의 데이터들만 DataFrame으로 변환될 수 있음

In [ ]:
#1차원 데이터를 이용해 DataFrame 생성

import numpy as np

col_name1=['col']
list1=[1,2,3]
array1=np.array(list1)
print('array1 shape : ',array1.shape)

#리스트를 이용해 DataFrame 생성
df_list1=pd.DataFrame(list1,columns=col_name1)
print('1차원 리스트로 만든 DataFrame:\n',df_list1)

#넘파이 ndarray를 이용해 DataFrame 생성
df_array1=pd.DataFrame(array1,columns=col_name1)
print('1차원 ndarray로 만든 DataFrame:\n',df_array1)


array1 shape :  (3,)


NameError: ignored

> 1차원 형태의 데이터를 기반으로 DataFrame을 생성하므로 칼럼명이 한 개만 필요함

In [ ]:
#2차원 형태의 데이터를 기반으로 DataFrame 생성

#3개의 칼럼명 필요
col_name2=['col1','col2','col3']

#2행 3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환
list2=[[1,2,3],[11,12,13]]
array2=np.array(list2)
print('array2.shape: ',array2.shape)

df_list2=pd.DataFrame(list2,columns=col_name2)
print('2차원 리스트로 만든 DataFrame : ', df_list2)
df_array2=pd.DataFrame(array2,columns=col_name2)
print('2차원 ndarray로 만든 DataFrame :\n',df_array2)

array2.shape:  (2, 3)
2차원 리스트로 만든 DataFrame :     col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame :
    col1  col2  col3
0     1     2     3
1    11    12    13


  * 딕셔너리를 DataFrame으로 변환 시에는 딕셔너리의 Key는 칼럼명으로, 딕셔너리의 Value는 키에 해당하는 칼럼 데이터로 변환됨
    * Key의 경우는 문자열, 값의 경우 리스트(또는 ndarray) 형태로 딕셔너리를 구성함

In [ ]:
# 딕셔너리를 DataFrame으로 변환

#key는 문자열 칼럼명으로 매핑, Value는 리스트 형(또는 ndarray) 칼럼 데이터로 매핑
dict={'col1':[1,11], 'col2':[2,22],'col3':[3,33]}
df_dict=pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame:\n', df_dict)

딕셔너리로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    22    33


> Key 값은 칼럼명, 각 Value는 각 칼럼 데이터로 매핑됨

####(2) DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기

많은 머신러닝 패키지가 기본 데이터 형으로 넘파이 ndarray를 사용하기 때문에 데이터 핸들링은 DataFrame으로 하더라도 머신러닝 패키지의 입력 인자 등에 적용하기 위해 다시 넘파이 ndarray로 변환하는 경우가 빈번하게 발생함.



In [ ]:
# DataFrame을 ndarray로 변환
array3=df_dict.values
print('df_dict.values 타입 : ', type(array3), 'df_dict.values shape : ',df_dict.shape)
print(array3)

df_dict.values 타입 :  <class 'numpy.ndarray'> df_dict.values shape :  (2, 3)
[[ 1  2  3]
 [11 22 33]]


* DataFrame.tolist() : DataFrame => list
* DataFrame.to_dict() : DataFrame => 딕셔너리
  * 인자로 'list'를 입력하면 딕셔너리의 값이 리스트형으로 반환됨

In [ ]:
#DataFrame을 리스트로 변환
list3=df_dict.values.tolist()
print('df_dict.values.tolist 타입 : ', type(list3))
print(list3)

#DataFrame을 딕셔너리로 변환
dict3=df_dict.to_dict('list')
print('\n df_dict.to_dict() 타입 : ',type(dict3))
print(dict3)

df_dict.values.tolist 타입 :  <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_dict() 타입 :  <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


###3) DataFrame의 칼럼 데이터 세트 생성과 수정

DataFrame의 칼럼 데이터 세트 생성과 수정 역시 [ ]연산자를 이용해 수행함

In [ ]:
# Age_0 칼럼 생성 및 0 할당
titanic_df['Age_0']=0
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

> 새로운 칼럼명 'Age_0'으로 모든 데이터 값이 0으로 할당된 Series가 기존 DataFrame에 추가됨

In [ ]:
# 기존 Series의 데이터를 이용해 새로운 칼럼 Series 생성
titanic_df['Age_by_10']=titanic_df['Age']*10
titanic_df['Family_No']=titanic_df['SibSp']+titanic_df['Parch']+1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

> 기존 칼럼 Series를 가공해 새로운 칼럼 Series인 Age_by_10과 Famil_No가 새롭게 DataFrame에 추가됨

In [ ]:
# DataFrame 내의 기존 칼럼 값 일괄 업데이트
titanic_df['Age_by_10']=titanic_df['Age_by_10']+100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

###4) DataFrame 데이터 삭제

* DataFrame.drop(labels=None, axis=0, index=None, colums=None, level=None, inplace=False, errors='raise')
  * axis 값에 따라 특정 칼럼 또는 특정 행을 drop
    * axis 0 은 로우 방향 축, axis 1은 칼럼 방향 축
    * aixs=1 => 칼럼을 drop
    * axis=0 => 특정 로우를 drop => label에 오는 값을 index로 간주 (이상치 제거에 사용)
  * inplace
    * False : 원본 DataFrame은 유지하고 드롭된 DataFrame을 새롭게 객체 벼수로 받고 싶은 경우 (Default)
    * True : 원본 DataFrame에 드롭된 결과를 적용할 경우 -> 반환 값은 None (반환 값을 다시 자기 자신에게 할당하면 안됨)
  * labels : 여러개의 칼럼을 삭제하려면 리스트 형태로 삭제하고자 하는 칼럼 명을 입력해 labels 파라미터로 입력하면 됨


In [ ]:
# 칼럼 삭제
titanic_drop_df=titanic_df.drop('Age_0',axis=1)
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

In [ ]:
drop_result=titanic_df.drop(['Age_0','Age_by_10','Family_No'],axis=1,inplace=True)
print(' inplace=True 로 drop 후 반환된 값: ', drop_result)
titanic_df.head(3) 

 inplace=True 로 drop 후 반환된 값:  None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [ ]:
# axis=0 설정
pd.set_option('display.width',1000)
pd.set_option('display.max_colwidth',15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))
titanic_df.drop([0,1,2],axis=0,inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


###5) Index 객체

판다스의 Index 객체는 DataFrame, Series의 레코드를 고유하게 식별하는 객체
* DataFrame, Series에서 Index 객체만 추출하려면 DataFrame.index 또는 Series.index 속성을 이용

In [ ]:
# 인덱스 객체 추출

# 원본 파일 다시 로딩
titanic_df=pd.read_csv('titanic_train.csv')

# Index 객체 추출
indexes=titanic_df.index
print(indexes)

#Index 객체를 실제 값 array로 변환
print('Index 객체 array 값:\n', indexes.values)

RangeIndex(start=0, stop=891, step=1)
Index 객체 array 값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 23

* Index 객체는 식별성 데이터를 1차원 array로 가지고 있음.
* ndarray와 유사하게 단일 값 반환 및 슬라이싱 가능

In [ ]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


* 한 번 만들어진 DataFrame 및 Series의 Index 객체는 함부로 변경할 수 없음

In [ ]:
# 인덱스 변경 불가
indexes[0]=5

TypeError: ignored

* Series 객체는 Index 객체를 포함하지만 Series 객체에 연산 함수를 적용할 때 Index는 연산에서 제외됨
* index는 오직 식별용으로만 사용됨



In [ ]:
series_fair=titanic_df["Fare"]
print('Fair Series max 값 : ', series_fair.max())
print('Fair Series sum 값 : ', series_fair.sum())
print('sum() Fair Series: ', sum(series_fair))
print('Fair Series + 3:\n', (series_fair+3).head(3))

Fair Series max 값 :  512.3292
Fair Series sum 값 :  28693.9493
sum() Fair Series:  28693.949299999967
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


* DataFrame.reset_index() 

: DataFrame 및 Series에 reset_index() 메서드를 수행하면 새롭게 인덱스를 연속 숫자 형으로 할당하며 기존 인덱스는 'index'라는 새로운 칼럼 명으로 추가됨
  * 인덱스가 연속된 int 숫자형 데이터가 아닐 경우에 다시 이를 연속 int 숫자형 데이터로 만들기 위해 사용
  * drop=True 로 설정 시 기존 인덱스가 새로운 칼럼명으로 추가되지 않고 삭제됨. > 새로운 칼럼이 추가되지 않으므로 그대로 Series로 유지됨

In [ ]:
titanic_reset_df=titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

In [ ]:
#Series와 value_counts()를 소개하는 예제에서 'Pclass' 칼럼 Series의 vlaue_counts()를 수행하면 'Pclass'고유 값이 식별자 인덱스 역할을 함. 이보다는 연속 숫자형 인덱스가 고유 식별자로 더 적합하기 때문에 reset_index()사용

print('### before reset_index ###')
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입: ', type(value_counts))
new_value_counts=value_counts.reset_index(inplace=False)
print("### after reset_index ###")
print(new_value_counts)
print('new_value_counts 객체 변수 타입: ',type(new_value_counts))

### before reset_index ###
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입:  <class 'pandas.core.series.Series'>
### after reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입:  <class 'pandas.core.frame.DataFrame'>


> 새롭게 연속 숫자형 인덱스가 만들어지고 기존 인덱스는 index 칼럼명으로 추가되면서 DataFrame으로 변환됨

###6) 데이터 셀렉션 및 필터링

ix[], iloc[], loc[] 연산자를 통해 연산자 내 단일 값 추출, 슬라이싱, 팬시 인덱싱, 불린 인덱싱 진행

####(1) DataFrame의 [] 연산자
* DataFrame 바로 뒤에 있는 '[]'는 칼럼 지정 연산자 -> [] 안에는 칼럼 명 문자(또는 칼럼 명의 리스트 객체), 또는 인덱스로 변환 가능한 표현식만 가능


In [11]:
print('단일 칼럼 데이터 추출:\n', titanic_df['Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출 :\n', titanic_df[['Survived','Pclass']].head(3))
print('[]안에 숫자 index는 Key Error 오류 발생:\n', titanic_df[0])

단일 칼럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출 :
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: ignored

* 판다스의 인덱스 형태로 변환 가능한 표현식은 []내에 입력할 수 있음

In [12]:
titanic_df[0:2]

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

* 불린 인덱싱 표현도 가능

In [14]:
titanic_df[titanic_df['Pclass']==3].head(3)

PassengerId  Survived  Pclass                      Name     Sex   Age  \
0            1         0       3   Braund, Mr. Owen Harris    male  22.0   
2            3         1       3    Heikkinen, Miss. Laina  female  26.0   
4            5         0       3  Allen, Mr. William Henry    male  35.0   

   SibSp  Parch            Ticket   Fare Cabin Embarked  
0      1      0         A/5 21171  7.250   NaN        S  
2      0      0  STON/O2. 3101282  7.925   NaN        S  
4      0      0            373450  8.050   NaN        S

* Data Frame 뒤의 [] 연산자 입력 인자
  * DataFrame 바로 뒤의 [] 연산자는 넘파이의 []와 Series의 []와 다름
  * DataFrame 바로 뒤의 [] 내 입력 값은 칼럼명(또는 칼럼의 리스트)을 지정해 칼럼 지정 연산에 사용하거나 불린 인덱스 용도로만 사용해야 함
  * DataFrame[0:2]와 같은 슬라이싱 연산으로 데이터를 추출하는 방법은 사용하지 않는 것이 좋음
  

#### (2)DataFrame ix[] 연산자

* ix[] : 행 위치에 인덱스 값, 열 위치에 칼럼 명(혹은 열 위치 값)을 지정해 원하는 데이터를 추출 
  * ix[0,'Pclass']와 같이 행 위치 지정으로 인덱스 값 0, 열 위치 지정으로 칼럼 명인 'Pclass'를 입력해 원하는 위치의 데이터를 추출할 수 있음 
  * 칼럼 명칭 기반 인덱싱,칼럼 위치 기반 인덱싱 모두 가능


In [15]:
print('칼럼 위치 기반 인덱싱 데이터 추출: ', titanic_df.ix[0,2])
print('칼럼 명 기반 인덱싱 데이터 추출:', titanic_df.ix[0,'Pclass'])

AttributeError: ignored

In [18]:
data={'Name':['Chulmin','Eunkyung','Jinwoong','Soobeom'],
      'Year':[2011,2016,2015,2015],
      'Gender':['Male','Female','Male','Male']}
data_df=pd.DataFrame(data, index=['one','two','three','four'])
data_df

Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

In [19]:
data_df.ix[0,0]

AttributeError: ignored

In [20]:
data_df.ix['one',0]

AttributeError: ignored

#### (3) 명칭 기반 인덱싱과 위치 기반 인덱싱의 구분

* 명칭 기반 인덱싱 : 칼럼의 명칭을 기반으로 위치를 지정하는 방식
* 위치 기반 인덱싱 : 0을 출발점으로 하는 가로축, 세로축 좌표 기반의 행과 열 위치를 기반으로 데이터를 지정

In [25]:
#data_df를 reset_index()로 새로운 숫자형 인덱스를 생성
data_df_reset=data_df.reset_index()
data_df_reset=data_df_reset.rename(columns={'index':'old_index'})

#인덱스값에 1을 더해서 1부터 시작하는 새로운 인덱스값 생성
data_df_reset.index=data_df_reset.index+1
data_df_reset

data_df_reset.ix[1,1]

AttributeError: ignored

In [ ]:
> ix[]의 경우 행과 열 위치에 명칭과 위치 기반 인덱싱 모두를 허용함. 그런데 일반적으로 행 위치에 적용되는 인덱스 값과 위치 기반 인덱싱이 integer형일 때 코드 작성에 혼란을 초래할 우려가 있음

####(4) DataFrame iloc[] 연산자

* iloc[]은 위치 기반 인덱싱만 허용 -> 행과 열 값으로 integer 또는 integer 형의 슬라이싱, 팬시 리스트 값을 입력해야 함
* DataFrame의 첫번째 행, 첫번째 열의 데이터를 추출하려면 data_df.iloc[0,0]과 같이 행과 열 위치에 위치 기반 인덱싱 값을 입력하면 됨
* 슬라이싱과 팬시 인덱싱은 제공하나 명확한 위치 기반 인덱싱이 사용되어야 하는 제약으로 인해 불린 인덱싱은 제공하지 않음

In [29]:
data_df.iloc[0,0]

#명칭 입력시 오류 발생
#data_df.iloc[0,'Name']

#문자 열 인덱스를 행 위치에 입력시 오류 발생
#data_df.iloc['one',0]

ValueError: ignored

In [34]:
# 앞의 ix[]예제에서 data_df를 reset_index()로 1부터 시작하는 새로운 정수형 인덱스값으로 변경한 data_df_reset에 iloc[0,1]을 적용

data_df_reset.iloc[0,1]

old_index      Name  Year  Gender
1       one   Chulmin  2011    Male
2       two  Eunkyung  2016  Female
3     three  Jinwoong  2015    Male

> iloc은 명확히 위치 기반 인덱싱이므로 첫번째 행(0)의 두번째 열(1)의 위치에 있는 값을 성공적으로 반환함

#### (5) DataFrame loc[] 연산자

* loc[]은 명칭 기반으로 데이터를 추출
  * 행위치에는 DataFrame index 값을, 열 위치에는 칼럼 명을 입력
  * index가 숫자형일 수 있기 때문에 명칭 기반이라고 무조건 문자열을 입력한다는 선입견을 가져서는 안됨.


In [35]:
# 인덱스 값이 'one'인 행의 칼럼 명이 'name'인 데이터를 추출

data_df.loc['one','Name']

'Chulmin'

In [38]:
data_df_reset.loc[1,'Name']

#다음 코드는 오류 발생(인덱스 값이 0인 행이 없음)
data_df_reset.loc[0,'Name']

KeyError: ignored

* loc에 슬라이싱 적용 시 종료 값까지 포함함
  * 명칭 기반 인덱싱이기 때문에, 인덱스가 숫자가 아닌 경우에는 -1을 할 수 없으므로 종료값까지 포함
  

In [41]:
#print('명칭 기반 ix slicing\n', data_df.ix['one':'two','Name'],'\n')
print('위치 기반 ix slicing\n',data_df.iloc[0:1,0],'\n')
print('명칭 기반 loc slicing\n',data_df.loc['one':'two','Name'])

위치 기반 ix slicing
 one    Chulmin
Name: Name, dtype: object 

명칭 기반 loc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


> 명칭 기반은 두 개의 행을 반환, 위치 기반은 한 개의 행을 반환


In [42]:
print(data_df_reset.loc[1:2,'Name'])

1     Chulmin
2    Eunkyung
Name: Name, dtype: object


> 인덱스 값에 따른 명칭 기반 인덱싱이므로 인덱스가 1,2인 2개의 데이터를 반환

####(6) 불린 인덱싱

* [],ix[],loc[]에서 공통으로 지원
* iloc[]은 정수형 값이 아닌 불린 값에 대해서는 지원하지 않기 때문에 불린 인덱싱이 지원되지 않음


* [] 을 이용한 불린 인덱싱
  * [] 내에 불린 인덱싱을 적용하면 반환 되는 객체가 DataFrame이므로 원하는 칼럼 명만 별도로 추출할 수 있음

In [43]:
# []연산자로 60세 이상인 타이타닉 데이터 추출
titanic_df=pd.read_csv('titanic_train.csv')
titanic_boolean=titanic_df[titanic_df['Age']>60]
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass                                       Name  \
33            34         0       2                      Wheadon, Mr. Edward H   
54            55         0       1             Ostby, Mr. Engelhart Cornelius   
96            97         0       1                  Goldschmidt, Mr. George B   
116          117         0       3                       Connors, Mr. Patrick   
170          171         0       1                  Van der hoef, Mr. Wyckoff   
252          253         0       1                  Stead, Mr. William Thomas   
275          276         1       1          Andrews, Miss. Kornelia Theodosia   
280          281         0       3                           Duane, Mr. Frank   
326          327         0       3                  Nysveen, Mr. Johan Hansen   
438          439         0       1                          Fortune, Mr. Mark   
456          457         0       1                  Millet, Mr. Francis Davis   
483          484         1       3                     Turkula, Mrs. (Hedwig)   
493          494         0       1                    Artagaveytia, Mr. Ramon   
545          546         0       1               Nicholson, Mr. Arthur Ernest   
555          556         0       1                         Wright, Mr. George   
570          571         1       2                         Harris, Mr. George   
625          626         0       1                      Sutton, Mr. Frederick   
630          631         1       1       Barkworth, Mr. Algernon Henry Wilson   
672          673         0       2                Mitchell, Mr. Henry Michael   
745          746         0       1               Crosby, Capt. Edward Gifford   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   
851          852         0       3                        Svensson, Mr. Johan   

        Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked  
33     male  66.0      0      0   C.A. 24579   10.5000          NaN        S  
54     male  65.0      0      1       113509   61.9792          B30        C  
96     male  71.0      0      0     PC 17754   34.6542           A5        C  
116    male  70.5      0      0       370369    7.7500          NaN        Q  
170    male  61.0      0      0       111240   33.5000          B19        S  
252    male  62.0      0      0       113514   26.5500          C87        S  
275  female  63.0      1      0        13502   77.9583           D7        S  
280    male  65.0      0      0       336439    7.7500          NaN        Q  
326    male  61.0      0      0       345364    6.2375          NaN        S  
438    male  64.0      1      4        19950  263.0000  C23 C25 C27        S  
456    male  65.0      0      0        13509   26.5500          E38        S  
483  female  63.0      0      0         4134    9.5875          NaN        S  
493    male  71.0      0      0     PC 17609   49.5042          NaN        C  
545    male  64.0      0      0          693   26.0000          NaN        S  
555    male  62.0      0      0       113807   26.5500          NaN        S  
570    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S  
625    male  61.0      0      0        36963   32.3208          D50        S  
630    male  80.0      0      0        27042   30.0000          A23        S  
672    male  70.0      0      0   C.A. 24580   10.5000          NaN        S  
745    male  70.0      1      1    WE/P 5735   71.0000          B22        S  
829  female  62.0      0      0       113572   80.0000          B28      NaN  
851    male  74.0      0      0       347060    7.7750          NaN        S

> titanic_Df[titanic_df['Age'>60]] 은 'Age' 칼럼이 60보다 큰 데이터를 모두 반환함

> 반환된 titanic_boolean 객체의 타입은 DataFrame


In [46]:
titanic_df[titanic_df['Age']>60][['Name','Age']].head(3)

Name   Age
33           Wheadon, Mr. Edward H  66.0
54  Ostby, Mr. Engelhart Cornelius  65.0
96       Goldschmidt, Mr. George B  71.0

 * loc[]을 이용한 불린 인덱싱

In [50]:
titanic_df.loc[titanic_df['Age']>60,['Name','Age']].head(3)



Name   Age
33           Wheadon, Mr. Edward H  66.0
54  Ostby, Mr. Engelhart Cornelius  65.0
96       Goldschmidt, Mr. George B  71.0

In [49]:
#여러 개의 복합 조건 설정
titanic_df[(titanic_df['Age']>60)&(titanic_df['Pclass']==1)& (titanic_df['Sex']=='female')]

PassengerId  Survived  Pclass                                       Name  \
275          276         1       1          Andrews, Miss. Kornelia Theodosia   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
275  female  63.0      1      0   13502  77.9583    D7        S  
829  female  62.0      0      0  113572  80.0000   B28      NaN

In [52]:
#개별 조건을 변수에 할당하고 이들 변수를 결합해서 불린 인덱싱을 수행

cond1=titanic_df['Age']>60
cond2=titanic_df['Pclass']==1
cond3=titanic_df['Sex']=='female'
titanic_df[cond1& cond2& cond3]

PassengerId  Survived  Pclass                                       Name  \
275          276         1       1          Andrews, Miss. Kornelia Theodosia   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
275  female  63.0      1      0   13502  77.9583    D7        S  
829  female  62.0      0      0  113572  80.0000   B28      NaN

###7) 정렬, Aggregation 함수, GroupBy 적용

####(1) DataFrame, Series의 정렬 - sort_values()

* sort_values()
  * by로 특정 칼럼을 입력하면 해당 칼럼으로 정렬을 수행
  * ascending=True : 오름차순 정렬 (Default) / ascending =False : 내림차순 정렬
  * inplace=False : sort_values()호출한 DataFrame 그대로 유지, 정렬된 DataFrame을 반환 (Default) / inplace = True : 호출한 DataFrame에 정렬된 결과 덮어쓰기 

In [53]:
# titanic_df를 Name 칼럼으로 오름차순 정렬해 반환
titanic_sorted = titanic_df.sort_values(by=['Name'])
titanic_sorted.head(3)

PassengerId  Survived  Pclass                              Name     Sex  \
845          846         0       3               Abbing, Mr. Anthony    male   
746          747         0       3       Abbott, Mr. Rossmore Edward    male   
279          280         1       3  Abbott, Mrs. Stanton (Rosa Hunt)  female   

      Age  SibSp  Parch     Ticket   Fare Cabin Embarked  
845  42.0      0      0  C.A. 5547   7.55   NaN        S  
746  16.0      1      1  C.A. 2673  20.25   NaN        S  
279  35.0      1      1  C.A. 2673  20.25   NaN        S

In [54]:
#여러 개의 칼럼으로 정렬
titanic_sorted=titanic_df.sort_values(by=['Pclass','Name'],ascending=False)
titanic_sorted.head(3)

PassengerId  Survived  Pclass                             Name   Sex  \
868          869         0       3      van Melkebeke, Mr. Philemon  male   
153          154         0       3  van Billiard, Mr. Austin Blyler  male   
282          283         0       3        de Pelsmaeker, Mr. Alfons  male   

      Age  SibSp  Parch    Ticket  Fare Cabin Embarked  
868   NaN      0      0    345777   9.5   NaN        S  
153  40.5      0      2  A/5. 851  14.5   NaN        S  
282  16.0      0      0    345778   9.5   NaN        S

####(2) Aggregation 함수 적용
  * DataFrame에서 바로 aggreation을 호출할 경우 모든 칼럼에 해당 aggregation을 적용함
  * 특정 칼럼에 aggregation 함수를 적용하기 위해서는 DataFrame에 대상 칼럼들만 추출해 aggregation을 적용하면 됨

In [55]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [56]:
titanic_df[['Age','Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

####(3) groupby() 적용

* groupby() : by에 칼럼을 입력하면 대상 칼럼으로 groupby 됨
  * DataFrame에 groupby()를 호출하면 DataFrameGroupBy라는 또다른 형태의 DataFrame을 반환함
  * groupby()를 호출해 반환된 결과에 aggregation 함수를 호출하면 groupby() 대상 칼럼을 제외한 모든 칼럼에 해당 aggregation 함수를 적용함

In [57]:
titanic_groupby=titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [58]:
titanic_groupby=titanic_df.groupby('Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
Pclass                                                                      
1               216       216   216  216  186    216    216     216   216   
2               184       184   184  184  173    184    184     184   184   
3               491       491   491  491  355    491    491     491   491   

        Cabin  Embarked  
Pclass                   
1         176       214  
2          16       184  
3          12       491

  * DataFrame의 groupby()에 특정 칼럼만 aggregation 함수를 적용하려면 groupby()로 반환된 DataFrameGroupBy 객체에 해당 칼럼을 필터링 한 뒤 aggregation 함수르 적용해야 함

In [59]:
titanic_groupby=titanic_df.groupby('Pclass')[['PassengerId','Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

* 서로 다른 aggregation 함수를 적용할 경우 적용하려는 여러 개의 aggregation 함수 명을 DataFrameGroupBy 객체의 agg()내에 인자로 입력해서 사용

In [60]:
titanic_df.groupby('Pclass')['Age'].agg([max,min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

###8) 결손 데이터 처리하기
* 결손 데이터는 넘파이의 Nan으로 표기


####(1) isna()로 결손 데이터 여부 확인
* isna() : 결손 데이터 여부 확인

In [61]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0        False     False   False  False  False  False  False  False   False   
1        False     False   False  False  False  False  False  False   False   
2        False     False   False  False  False  False  False  False   False   

    Fare  Cabin  Embarked  
0  False   True     False  
1  False  False     False  
2  False   True     False

* 결손 데이터의 수는 isna()와 sum()을 사용하여 구함

In [62]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

#### (2) fillna()로 결손 데이터 대체하기
* fillna()를 이용해 반환 값을 다시 받거나 inplace=True 파라미터를 fillna()에 추가해야 실제 데이터 세트 값이 변경됨

In [63]:
titanic_df['Cabin']=titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  C000        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  C000        S

In [64]:
titanic_df['Age']=titanic_df['Age'].fillna(titanic_df['Age'].mean())
titanic_df['Embarked']=titanic_df['Embarked'].fillna('S')
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

####(3) apply lambda 식으로 데이터 가공
* 판다스는 apply 함수에 lambda 식을 결합해 DataFrame이나 Series의 레코드별로 데이터를 가공하는 기능을 제공함

* lambda식

In [66]:
# 입력값의 제곱 값을 구해서 반환하는 get_square(a)
def get_square(a):
  return a**2

print('3의 제곱은 ', get_square(3))

3의 제곱은  9


* lambda는 함수의 선언과 함수 내의 처리를 한 줄의 식으로 쉽게 변환하는 식
  * : 왼쪽에는 입력 인자, : 오른쪾에는 반환 값

In [67]:
# 위의 get_square를 lambda식으로 변환
lambda_square=lambda x:x**2
print('3의 제곱은', lambda_square(3))

3의 제곱은 9


* lambda 식 사용 시 여러 개의 값을 입력 인자로 사용해야 할 경우 map()함수를 결합하여 사용


In [68]:
a=[1,2,3]
squares=map(lambda x:x**2,a)
list(squares)

[1, 4, 9]

In [70]:
#lambda 식을 이용한 데이터 가공

titanic_df['Name_len']=titanic_df['Name'].apply(lambda x: len(x))
titanic_df[['Name','Name_len']].head(3)

Name  Name_len
0                            Braund, Mr. Owen Harris        23
1  Cumings, Mrs. John Bradley (Florence Briggs Th...        51
2                             Heikkinen, Miss. Laina        22

In [72]:
#lambda에서 ifelse 식을 사용해 조금 더 복잡한 데이터 가공

#나이가 15세 미만이면 Child, 그렇지 않으면 Adult로 구분하는 새로운 칼럼 Child_Adult
titanic_df['Child_Adult']=titanic_df['Age'].apply(lambda x:'Child' if x<=15 else 'Adult')
titanic_df[['Age','Child_Adult']].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

* lambda 식에서 if절 사용 시 if 식보다 반환 값을 먼저 서술해야 함
* else 식의 경우 else가 먼저 나오고 반환값이 나중에 나옴
* if,else만 지원하고 else if는 지원하지 않음 > elseif를 사용하려면 else절을 ()로 내포해 () 내에서 다시 if else를 적용

In [73]:
# 만 15세 이하는 Child, 15~60 Adult, 61세 이상은 Elerly로 분류하는 Age_Cat 칼럼
titanic_df['Age_Cat']=titanic_df['Age'].apply(lambda x: 'Child' if x<=15 else ('Adult' if x<=60 else 'Elderly'))
titanic_df['Age_Cat'].value_counts()

Adult      786
Child       83
Elderly     22
Name: Age_Cat, dtype: int64

* else if를 계속 내포해서 쓰는것이 부담스러울 경우 함수를 생성하여 사용하는 것이 편함

In [79]:
# 나이에 따라 더 세분화 된 분류
def get_cartegory(age):
  cat=''
  if age<=5 : cat='Baby'
  elif age<=12: cat='Child'
  elif age<=18 : cat='Teenager'
  elif age<=25 : cat='Student'
  elif age<=35 : cat='Young Adult'
  elif age<=60 : cat='Adult'
  else : cat='Elderly'
  return cat
  
  #lambda 식에 get_cartegory() 함수를 반환값으로 지정
  # get_cartegory(x)는 입력값으로 'Age'칼럼을 받아서 해당하는 cat 반환
titanic_df['Age_Cat']=titanic_df['Age'].apply(lambda x: get_cartegory(x))
titanic_df[['Age','Age_Cat']].head()

Age      Age_Cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult

##__5. 정리__

넘파이와 판다스는 머신러닝 학습을 위해 매우 중요한 요소

넘파이와 판다스는 머신러닝 프로그램을 직접 만들면서 이해하고, 모르는 부분들을 인터넷 검색 등으로 찾아보는 것이 학습을 위한 제일 빠른 방법

실제 ML 모델을 생성하고 예측을 수행하는 데 있어서 ML 알고리즘이 차지하는 비중보다 데이터를 전처리하고 적절한 피처를 가공/추출하는 부분이 훨씬 많은 비중을 차지하게 됨. 따라서 바로 뒤에서 설명할 사이킷런에 대한 이해도 중요하지만, 넘파이, 판다스, 맷플롯립/시본 과 같이 파이썬 머신러닝 생태계를 이루고 있는 다양한 패키지에 대한 이해 역시 매우 중요함.